### PR TITLE
Use BeautifulSoup instead of lxml for MJML/HTML parsing

### LINKED ISSUE #6: HTML entities such us &nbsp; or &copy; are not supported
currently fails with:

```lxml.etree.XMLSyntaxError: Entity 'copy' not defined, line 853, column 19```

see also:
https://stackoverflow.com/questions/31210655/nbsp-text-not-handled-by-the-lxml

### LINKED ISSUE #10: cleanup "innerXML" functionality of mj-raw
The "innerXML" functionality for `mj-raw` is marked as "ugly" and should be cleaned up.

@barsch [wrote](https://github.com/FelixSchwarz/mjml-stub/pull/8#discussion_r561070039):
> there are multiple approaches to solve it:
> https://stackoverflow.com/questions/4624062/get-all-text-inside-a-tag-in-lxml
> https://stackoverflow.com/questions/11122397/how-do-i-get-the-whole-content-between-two-xml-tags-in-python
> I don't mind which one to use

### LINKED ISSUE #14: Orphans opening tags from HTML `<br>`
From Passport (mailjet's editor) I receive some  `<br>` HTML tags inside `mjml-text` blobs. In XHTML, those should actually be `<br/>` and it currently make the mjml-stub library choke.

1. Should I clean them up before using mjml-stub?
2. Can I try to fix the lib to accept them?

Note that mjml.io is able to work with those tags
https://mjml.io/try-it-live/gb2kuVgCdRn
The HTML is generated properly with br tags, there are no errors mentioned. Only the syntax highlighter gets lost.

I'd be in favor of doing 2, but don't want to start before your approval.

### LINKED ISSUE #19: Whitespace gets stripped between text & link inside <mj-text />
This appears to be a bug inside `mj-text` rendering. If I run this code:

```python
from mjml import mjml_to_html
print(mjml_to_html('<mjml><mj-body><mj-text>Hello <a href="#">world</a></mj-text></mj-body></mjml>')['html'].split('\n')[-3])
```

I get this output:

```html
    <div style=""><div style="font-family:Ubuntu, Helvetica, Arial, sans-serif;font-size:13px;line-height:1;text-align:left;color:#000000">Hello<a href="#">world</a></div></div>
```

The space between "Hello" and "<a" gets stripped during the parsing.

I haven’t been able to dive through the code enough, yet, but seems like the first children element inside this `mj-text` should have  `elem["tail"] == " "` https://github.com/FelixSchwarz/mjml-stub/blob/main/mjml/elements/mj_text.py#L104

### PR DESCRIPTION
**Pull request by caseyjhol**
- Fixes #6 
- Fixes #10 
- Fixes #14 
- Fixes #19

Some of the upstream alignment tests were failing. I passed the source MJML back into mjml-cli locally and rebuilt the HTML, and I then updated the "-expected.html" files (some style tags were missing). All tests then passed.

### CODE REVIEW

**Comment by FelixSchwarz** [comment]
That's quite an impressive pull request which could really fix a lot of issues, thank you so much. I didn't have time to really dig into this but I'll try to split my review into smaller pieces to understand what has changed and why.
Switching to `Beautiful Soup` might be a good solution to resolving the issues around the pretty strict `lxml` parsing and these issues are actually the single biggest core issue of this project right now so highly valuable.

I was wondering about failing upstream tests. So you are saying that after your `Beautiful Soup` migration some upstream tests were failing which did work before? And these tests started to pass once you regenerated the `...-expected.html` files?

- Which version of mjml (upstream) did you use?

I would love to structure the changes in a way that switching the parser does not require any changes to the tests.

What is your confidence in the correctness of your changes? Did you compare the HTML output manually against upstream HTML? Do you use the new code already in production? Once I got a chance to read up on your code I can plug it into a customer's test suite to ensure the new code did not change in unexpected ways.

Last but not least: Thank you for tackling this high-value issue. Looking forward to merging your code.

**Comment by caseyjhol** [comment]
Looks like it was only the HTML for the "missing features" tests that needed to be regenerated (after I moved them alongside our other test data). I used MJML v4.12.0 to generate - the HTML. These tests were already failing previously, of course, so it's possible issues with the HTML flew under the radar.

Related, I wonder if there's some way we could call mjml-cli from our tests to generate the expected HTML on the fly instead. It would be nice too, because we could then ensure we were always matching the same version of MJML in all of our tests.

File: mjml/core/api.py [hunk 1]
```diff
@@ -72,7 +71,7 @@ def getContent(self):
         # mjml's "email-templates" repo.
         if self.content is None:
             return ''
-        return self.content.strip()
+        return str(self.content).strip()
```
Comment (caseyjhol) [inline mjml/core/api.py:74]: Might not be necessary here to cast to a string, since this should already be one (I think this is a remnant from WIP).

**Comment by FelixSchwarz** [comment]
> These tests were already failing previously, of course, so it's possible issues with the HTML flew under the radar.

That makes sense.

> Related, I wonder if there's some way we could call mjml-cli from our tests to generate the expected HTML on the fly instead. It would be nice too, because we could then ensure we were always matching the same version of MJML in all of our tests.

Generating it on the fly could bring quite a few issues and makes the test suite more brittle but I have a script to regenerate all HTML files. I could add this to the git repo.

**Comment by FelixSchwarz** [comment]
Ok, I went through your changes and I think they are fine. I noticed that the `...-expected.html` in the `missing_functionality` were out of date so I updated them in a separate commit fb83576. Unfortunately github's web editor can not handle that.

Would you mind rebasing your commit on top of `main` + squashing everything into a single commit? Bonus points for explaining the motivation the rationale of that change (something about the tickets you mentioned in your initial comment) in the commit comment.

As a final validation I'd like to plug your changes into a customer's CI pipeline to see if anything breaks there. I'll do that on Monday. After that we're ready to merge :-)

**Comment by caseyjhol** [comment]
@FelixSchwarz Alright - should be all set!

**Comment by FelixSchwarz** [comment]
Everything looked fine in our CI pipeline so I think there should not be too many surprises :-)

One thing which might be affected is that BeautifulSoup is more leniant when it comes to XML syntax so maybe we should add some (optional) XSD validation to ensure the inputs not completely off-base.

With your changes now in main there are enough new features/bug fixes to warrant a release. I plan on doing that shortly once I completed a few tasks:

- #24 
- #23 
- #22 

Thank you very much, the BeautifulSoup changes much less intrusive than I expected and really solves a lot of issues I had with lxml.

**Pull request merged**